genapi/genapi.go

```diff
@@ -0,0 +1,361 @@
+	"github.com/mediocregopher/lever"
+	"github.com/mediocregopher/radix.v2/util"
+	"github.com/mediocregopher/skyapi/client"
+	"gopkg.in/mgo.v2"
+)
+
+// MongoInfo contains information needed by the api to interact with a mongo
+// backend, and also houses the connection to that backend (which can be
+// interacted with through its methods)
+type MongoInfo struct {
+	// The name of the mongo database this app should use. In TestMode this will
+	// always be overwritten to "test_<DBName>"
+	DBName string
+
+	session *mgo.Session
+}
+
+// WithDB is similar to mgoutil.SessionHelper's WithDB, see those docs for more
+// details
+func (m MongoInfo) WithDB(fn func(*mgo.Database)) {
+	mgoutil.SessionHelper{
+		Session: m.session,
+		DB:      m.DBName,
+	}.WithDB(fn)
+}
+
+// WithColl is similar to mgoutil.SessionHelper's WithColl, see those docs for
+// more details
+func (m MongoInfo) WithColl(collName string, fn func(*mgo.Collection)) {
+	mgoutil.SessionHelper{
+		Session: m.session,
+		DB:      m.DBName,
+		Coll:    collName,
+	}.WithColl(fn)
+}
+
+// CollSH returns an mgoutil.SessionHelper for a collection of the given name
+func (m MongoInfo) CollSH(collName string) mgoutil.SessionHelper {
+	return mgoutil.SessionHelper{
+		Session: m.session,
+		DB:      m.DBName,
+		Coll:    collName,
+	}
+}
+
+func (m MongoInfo) isZero() bool {
+	return m == MongoInfo{}
+}
+
+// RedisInfo is used to tell the api to interact with a redis backend, and also
+// houses the connection to that backend. If the redis backend is a cluster
+// instance that whole cluster will be connected to
+type RedisInfo struct {
+	// Set to true to tell the api to connect to redis. In the future, if there
+	// are other fields which may be set, this may become optional
+	Use bool
+
+	// Populated by the api once a connection to redis is made, and can be used
+	// as such. Do not set manually.
+	util.Cmder
+}
+
+func (r RedisInfo) isZero() bool {
+	return r.Use
+}
+
+// GenAPI is a type used to handle most of the generic logic we always implement
+// when making an RPC API endpoint.
+//
+// The struct is initialized with whatever parameters are appropriate, and then
+// has either APIMode(), TestMode(), or CLIMode() called on it depending on the
+// intent. Fields are required to be filled in unless marked as optional in
+// their comment.
+type GenAPI struct {
+	// Name is the name of the api, as it will be identified on the command-line
+	// and in skyapi
+	Name string
+
+	// DefaultListenAddr is the address that will be listened on by default
+	DefaultListenAddr string
+
+	// Do not set. This will be automatically filled in when any of the run
+	// modes are called, and may be used after that point to retrieve parameter
+	// values.
+	*lever.Lever
+
+	// Optional. Additional lever.Param structs which can be included in the
+	// lever parsing
+	LeverParams []lever.Param
+
+	// Optional. If mongo is intended to be used as a backend, this should be
+	// filled in
+	MongoInfo
```

Comment:
Same thing with this being a pointer. It would also get rid of the `IsZero` stuff.